frontend/catalyst/ag_primitives.py

```diff
@@ -134,9 +319,95 @@ def converted_call(fn, *args, **kwargs):
 
             @functools.wraps(fn.func)
             def qnode_call_wrapper():
-                return tf_converted_call(fn.func, *args, **kwargs)
+                return tf_converted_call(fn.func, args, kwargs, caller_fn_scope, options)
 
             new_qnode = qml.QNode(qnode_call_wrapper, device=fn.device, diff_method=fn.diff_method)
             return new_qnode()
 
-        return tf_converted_call(fn, *args, **kwargs)
+        return tf_converted_call(fn, args, kwargs, caller_fn_scope, options)
+
+
+class CRange:
+    """Catalyst range object.
+
+    Can be passed to a Python for loop for native conversion to a for_loop call.
+    Otherwise this class behaves exactly like the Python range class.
+
+    Without this native conversion, all iteration targets in a Python for loop must be convertible
+    to arrays. For all other inputs the loop will be treated as a regular Python loop.
+    """
+
+    def __init__(self, start_stop, stop=None, step=None):
+        self._py_range = None
+        self._start = start_stop if stop is not None else 0
+        self._stop = stop if stop is not None else start_stop
+        self._step = step if step is not None else 1
+
+    def get_raw_range(self):
+        """Get the raw values defining this range: start, stop, step."""
+        return self._start, self._stop, self._step
+
+    @property
+    def py_range(self):
+        """Access the underlying Python range object. If it doesn't exist, create one."""
+        if self._py_range is None:
+            self._py_range = range(self._start, self._stop, self._step)
+        return self._py_range
+
+    # Interface of the Python range class.
+    # pylint: disable=missing-function-docstring
+
+    @property
+    def start(self) -> int:
+        return self.py_range.start
```

Comment:
It looks like these class properties/methods not tested

Reply:
That's true. I was thinking of marking this section `nocover`, because it would require testing the entire interface of the builtin `range` function, and all these lines do is forward the API to the builtin `range`.

Reply:
Would be possible to create this class by inheriting from range? It would remove the testing issue and make things cleaner and more robust to possible python changes.

Reply:
Unfortunately no, it is marked final.

Reply:
For now I've marked the `range` interface itself as `nocover` (for the above reasons), but I did add a unit test that checks the fallback of `CRange` to `range` is working correctly: https://github.com/PennyLaneAI/catalyst/pull/258/commits/5341c3abc09d5f87a539ee99783487e4596786de

Let me know if this is acceptable :)